cpp/src/slice2py/PythonUtil.cpp

```diff
@@ -2543,92 +2576,220 @@ namespace
+                // compilation.
+                if (runtimeImports.find(moduleName) != runtimeImports.end() &&
+                    find_if(
+                        processedFragments.begin(),
+                        processedFragments.end(),
+                        [&moduleName](const auto& fragment)
+                        { return fragment.moduleName == moduleName; }) == processedFragments.end() &&
+                    find(builtinModules.begin(), builtinModules.end(), moduleName) == builtinModules.end())
                 {
-                    // TODO skip type hints that are already imported as part of the runtime imports.
-                    out << nl << "from " << moduleName << " import " << name << " as "
-                        << (alias.empty() ? name : alias);
+                    unseenDependencies = true;
+                    break;
                 }
             }
+
+            if (!unseenDependencies)
+            {
+                // If we have already processed all the dependencies we can remove it from the runtime imports.
+                // add add it to the list of seen modules.
+                Python::BufferedOutput out;
+                writeImports(runtimeImports[fragment.moduleName], typingImports[fragment.moduleName], out);
+                out << sp;
+                out << fragment.code;
+                processedFragments.push_back(
+                    {.sliceFileName = fragment.sliceFileName,
+                     .moduleName = fragment.moduleName,
+                     .fileName = fragment.fileName,
+                     .isPackageIndex = false,
+                     .code = out.str()});
+                it = fragments.erase(it);
+            }
             else
             {
-                // If there are no definitions, we still need to import the module to ensure that the type hints
-                // are available.
-                out << nl << "import " << moduleName;
+                it++;
             }
         }
-        out.dec();
-    }
 
-    // Emit the code fragments for the unit.
-    for (const auto& fragment : codeVisitor.codeFragments())
-    {
-        string moduleName;
-        if (dynamic_pointer_cast<ClassDecl>(fragment.contained) ||
-            dynamic_pointer_cast<InterfaceDecl>(fragment.contained))
-        {
-            moduleName = getPythonModuleForForwardDeclaration(fragment.contained);
-        }
-        else
+        // If we didn't remove any module from the runtime imports, it means that we have a circular dependency between
+        // Slice files which is not allowed.
+        if (fragments.size() == fragmentsSize)
         {
-            moduleName = getPythonModuleForDefinition(fragment.contained);
+            assert(false);
+            throw runtime_error(
+                "Circular dependency detected in Slice files. Please check the imports and package definitions.");
```

Comment:
If I'm not mistaken the Slice parser already ensures this, and we should probably remove this extra check.